active-rfcs/0000-router-navigation-failures.md

```diff
@@ -0,0 +1,173 @@
+- Start Date: 2020-03-26
+- Target Major Version: Vue Router v4
+- Reference Issues: https://github.com/vuejs/vue-router/issues/2833, https://github.com/vuejs/vue-router/pull/3047, https://github.com/vuejs/vue-router/issues/2932, https://github.com/vuejs/vue-router/issues/2881
+- Implementation PR:
+
+# Summary
+
+- Explicitely define what a navigation failure is and how and where we can catch them.
+- Change when the Promised-based `router.push`(and `router.replace` by extension) resolves and rejects.
+- Make `router.push` consistent with `router.afterEach` and `router.onError`.
+
+# Basic example
+
+## `router.push`
+
+If there is an unhandled error or an error is passed to `next`:
+
+```js
+// any other navigation guard
+router.beforeEach((to, from, next) => {
+  next(new Error())
+  // or
+  throw new Error()
+  // or
+  return Promise.reject(new Error())
+})
+```
+
+Then the promise returned by `router.push` rejects:
+
+```js
+router.push('/url').catch(err => {
+  // ...
+})
+```
+
+In **all** other cases, the promise is resolved. We can know if the navigation failed or not by checking the resolved value:
+
+```js
+promise.push('/dashboard').then(failure => {
+  if (failure) {
+    failure instanceof Error // true
+    failure.type // NavigationFailure.canceled
+  }
+})
+```
+
+## `router.afterEach`
+
+It's the global equivalent of `router.push().then()`
+
+## `router.onError`
+
+It's the global equivalent of `router.push().catch()`
+
+# Motivation
+
+The current behavior of Vue Router regarding the promise returned by `push` is inconsistent with `router.afterEach` and `router.onError`. Ideally, we should be able to catch all succeeded and failed navigations globally and locally but we can only do it locally.
+
+- `onError` is only triggered on thrown errors and `next(new Error())`
+- `afterEach` is only called iif there is a navigation
+- `redirect` should behave the same as `next('/url')` in a Navigation guard
```

Comment:
Should it? From current [documentation](https://router.vuejs.org/guide/essentials/redirect-and-alias.html#redirect):
> Note that Navigation Guards are not applied on the route that redirects, only on its target. In the example below, adding a beforeEnter guard to the /a route would not have any effect.

As I understand, `afterEach` (as well as `beforeEach` and route's `beforeEnter') will not be called for redirect routes. Will this change?

Reply:
You are right, it's not exactly the same and it should be pointed out